### ISSUE TITLE
CI Android is failing

### ISSUE THREAD

**Issue by DominusExult** [exult/exult#241]
@ceckak if you have time could you take a look why the CI Android is failing?
The CI github action file is https://github.com/exult/exult/blob/master/.github/workflows/ci-android.yml

and fail is recorded at https://github.com/exult/exult/runs/8052304193?check_suite_focus=true#step:10:6621
Something something with libvorbis is failing. Maybe something to do with https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=247799 ?

**Comment by ceckak** [comment]
Thanks for getting started on the CI for Android.  I will have limited availability through the end of the summer, but will start digging into this as time permits.  First up, I need to do a rebase of my local tree with all the recent changes and confirm it still builds manually and then I can review that against the CI results you are seeing.

**Comment by DominusExult** [comment]
Thank you!

**Comment by DominusExult** [comment]
One recent problem I "fixed" by setting gradle to <7 (6.9.4 currently). If it is set to "current" (or >=7) you get the build error in the zipped log:
[configurebuild.zip](https://github.com/exult/exult/files/11009890/configurebuild.zip)

The older problem is indeed fixed by adding `-lm` to the makefile.am with this patch:
```
diff --git a/lib/Makefile.am b/lib/Makefile.am
index e22895e..b51d5df 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -20,11 +20,11 @@ libvorbis_la_LIBADD = @VORBIS_LIBS@ @OGG_LIBS@
 
 libvorbisfile_la_SOURCES = vorbisfile.c
 libvorbisfile_la_LDFLAGS = -no-undefined -version-info @VF_LIB_CURRENT@:@VF_LIB_REVISION@:@VF_LIB_AGE@
-libvorbisfile_la_LIBADD = libvorbis.la @OGG_LIBS@
+libvorbisfile_la_LIBADD = libvorbis.la @OGG_LIBS@ -lm
 
 libvorbisenc_la_SOURCES = vorbisenc.c 
 libvorbisenc_la_LDFLAGS = -no-undefined -version-info @VE_LIB_CURRENT@:@VE_LIB_REVISION@:@VE_LIB_AGE@
-libvorbisenc_la_LIBADD = libvorbis.la @OGG_LIBS@
+libvorbisenc_la_LIBADD = libvorbis.la @OGG_LIBS@ -lm
 
 EXTRA_PROGRAMS = barkmel tone psytune
 CLEANFILES = $(EXTRA_PROGRAMS)

```
Now I have to see about how to add two patches.
Currently investigating other problems (on the github actions image ubuntu it picks up alsa and timidity and these need to be disabled in configure as well).

**Comment by DominusExult** [comment]
Something is wrong with Alsa... when I explicitly disable it and it checks for Alsa cmake responds that it can't find asoundlib.h and libasound.
But then in the synopsis it suddenly states it is enabled and then compiling fails

```
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : checking for alsa/asoundlib.h... no
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : checking for alsa/asoundlib.h... no
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : checking if we want to use ALSA midi... no; libasound not found
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : checking if we want to use ALSA midi... no; libasound not found
....
*snip*
....
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : Audio :
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : Enable ALSA midi........... : yes
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : Enable ALSA midi........... : yes
....
*snip*
....
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : libtool: compile:  /usr/local/lib/android/sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi28-clang++ -std=c++17 -DHAVE_CONFIG_H -I. -I/home/runner/work/exultCIfork/exultCIfork/audio/midi_drivers -I../.. -I/home/runner/work/exultCIfork/exultCIfork/headers -I/home/runner/work/exultCIfork/exultCIfork/conf -I/home/runner/work/exultCIfork/exultCIfork -I/home/runner/work/exultCIfork/exultCIfork/audio -I/home/runner/work/exultCIfork/exultCIfork/files -I/home/runner/work/exultCIfork/exultCIfork/imagewin -I/home/runner/work/exultCIfork/exultCIfork/objs -I/home/runner/work/exultCIfork/exultCIfork/shapes -I/home/runner/work/exultCIfork/exultCIfork/android/app/.cxx/cmake/debug/armeabi-v7a/dependencies/install/include/SDL2 -I/home/runner/work/exultCIfork/exultCIfork/android/app/.cxx/cmake/debug/armeabi-v7a/dependencies/install/include -DMT32EMU_EXPORTS_TYPE=0 -DMT32EMU_API_TYPE=0 -Wall -Wextra -pedantic -Wc++14-compat -Wc++17-compat -Wcast-align -Wcast-qual -Wctor-dtor-privacy -Wdisabled-optimization -Wextra-semi -Wformat-nonliteral -Wformat-security -Wlogical-not-parentheses -Wmissing-include-dirs -Wnon-virtual-dtor -Wnull-dereference -Wold-style-cast -Woverloaded-virtual -Wredundant-decls -Wshift-negative-value -Wshift-overflow -Wtrigraphs -Wundef -Wuninitialized -Wwrite-strings -Wzero-as-null-pointer-constant -Wabsolute-value -Wdeprecated-register -Wmismatched-tags -Wunused-private-field -O0 -fno-limit-debug-info -MT MT32EmuMidiDriver.lo -MD -MP -MF .deps/MT32EmuMidiDriver.Tpo -c /home/runner/work/exultCIfork/exultCIfork/audio/midi_drivers/MT32EmuMidiDriver.cpp -o MT32EmuMidiDriver.o >/dev/null 2>&1
10241
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : libtool:   error: 'ALSAMidiDriver.lo' is not a valid libtool object
10242

10243
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : libtool:   error: 'ALSAMidiDriver.lo' is not a valid libtool object
10244
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[7]: *** [Makefile:560: libmididrv.la] Error 1
10245

10246
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[7]: *** [Makefile:560: libmididrv.la] Error 1
10247
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[6]: *** [Makefile:624: all-recursive] Error 1
10248

10249
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[6]: *** [Makefile:624: all-recursive] Error 1
10250
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[5]: *** [Makefile:594: all-recursive] Error 1
10251

10252
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[5]: *** [Makefile:594: all-recursive] Error 1
10253
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[4]: *** [Makefile:1728: all-recursive] Error 1
10254

10255
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[4]: *** [Makefile:1728: all-recursive] Error 1
10256
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[3]: *** [Makefile:917: all] Error 2
10257

10258
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : make[3]: *** [Makefile:917: all] Error 2
10259
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : ninja: build stopped: subcommand failed.
10260

10261
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : ninja: build stopped: subcommand failed.
10262
/home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : {"cookie":"","inReplyTo":"configure","message":"CMake Error at dependencies/build_dependency.cmake:21 (message):\n  CMake build failed for exult-1.9.0git: 1\nCall Stack (most recent call first):\n  dependencies/exult/CMakeLists.txt:2 (include)\n\n","title":"Error","type":"message"}
10263

10264
Task :app:generateJsonModelDebug in app Finished
10265

10266

10267
> Task :app:generateJsonModelDebug FAILED
10268
:app:generateJsonModelDebug (Thread[Daemon worker,5,main]) completed. Took 1 mins 11.503 secs.
10269
AAPT2 aapt2-4.1.2-6503028-linux Daemon #0: shutdown
10270
AAPT2 aapt2-4.1.2-6503028-linux Daemon #1: shutdown
10271

10272
FAILURE: Build failed with an exception.
10273
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
10274
Use '--warning-mode all' to show the individual deprecation warnings.
10275
See https://docs.gradle.org/6.9.4/userguide/command_line_interface.html#sec:command_line_warnings
10276
13 actionable tasks: 13 executed
10277

10278
* What went wrong:
10279
Execution failed for task ':app:generateJsonModelDebug'.
10280
> /home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : CMake Error at /home/runner/work/exultCIfork/exultCIfork/android/app/src/main/cpp/dependencies/build_dependency.cmake:21 (message):
10281
    CMake build failed for exult-1.9.0git: 1
10282
  Call Stack (most recent call first):
10283
    dependencies/exult/CMakeLists.txt:2 (include)
```

**Comment by DominusExult** [comment]
🤦‍♂️
The big culprit was that for the Android build to work one needs to build outside of the source folder. I didn‘t realize that even though it looked me right in the eyes when I figured out that the normal Exult build messed with the APK builds… 🤦‍♂️

Anyway still errors out but much later after building the two ARM arches when it gets to building the x86 arch of vorbis.

**Comment by ceckak** [comment]
Ah - that makes sense.  Sorry for not being super-responsive on this, and thanks for continue to work on it.  I've been trying to carve out a little time again to work on some updates, and can try to take a look at this also.  What's the easiest way for me to see the build errors and/or test the CI build myself?

**Comment by DominusExult** [comment]
Thank you! No reason to feel bad about it, time is precious and sometimes you just don't have spare time :)

To view the build error you can go to the "Actions" tab of the project https://github.com/exult/exult/actions and click through the red ci-android actions. But let me see how to set it up so it builds out of source as right now you can only see build errors related to being built in source (looking into that right now).
The easiest to test actions is probably to fork Exult and do some few changes before getting started:
- go to the Actions tab in your fork and enable the actions.
- disable all actions except for ci-android (click on an action and in the top right, next to the "Filter workflow runs", click on the ... menu and select disable)
- edit in the source /.github/workflows/ci-android.yml and delete the "notify" job or you will end up spamming our irc channel (and get fail report as the discord notify will not work)
- regard that fork as a throwaway fork as you will end up doing weird stuff just to test what happens to the workflow if you do this or that :)

**Comment by DominusExult** [comment]
I've got it working now :)

There were several problems:
- building out of source is imperative
- Gradle needs to be <v7 (currently 6.9.4)
- the Github action runner image needs to be Ubuntu 20.04. The "latest" uses 22.04 and either that Ubuntu version or something that is installed crashes compilation without a clear indication what is going wrong (see https://github.com/exult/exult/actions/runs/4466704272/jobs/7845223889). Curiously my Ubuntu 22.10 VM works correctly (with all the fixes mentioned here).
- in  f5b48926e I removed the Clang patch for Vorbis as I mistakenly believed that it wasn't necessary (the ARM builds worked without this). This was important for building the x86 build.

I added comments to Readme.md about Gradle and out of source building as well as some shortcuts to building the APK (disabling almost everything in configure, only making the data files and then switching to the APK build).

All that said, it would be great if you could upgrade the gradle files so it works with current Gradle.

It was an interesting experience, especially as I got to understand what the build system does for the APK :)

**Comment by DominusExult** [comment]
Another status update:
Gradle 7.x also works and I enabled this as well as giving out warnings and eliminated one source of warnings. Gradle 8 won't work, though.

I have general ideas about the build system. 
As it stands the only things the Android build needs from the Exult configure.ac is to check general build stuff, Gradle version, SDK and NDK versions. It doesn't actually need to check for the usual Exult dependencies as these get downloaded by the Android build. It doesn't even need a valid Cmake install (I *think*) as the Android build uses the Cmake included in the SDK (right)?
So I think a lot, especially the Exult dependencies can go into an `enable_android_apk = xno` test and then also force enable building the data files into a `enable_android_apk != xno` test (along with a --disable-exult).
And then, once configure has run its course, make should only run in ./files, ./data, ./android (in that order).
(No idea if that works for SDL2 , though, because of the acinclude.m4 things)

**Comment by Dragon-Baroque** [comment]
> No idea if that works for SDL2 , though, because of the acinclude.m4 things.

Precisely. On my personal source tree, I have eradicated the whole `m4/` directory as being either not used ( `gb_enable_lto` ) or outdated ( all the `ax_` ). But `acinclude.m4` remains with its `EXULT_CHECK_SDL` which I would like to replace by a normal `pkg-config "sdl2 >= 2.0.10"`, or whatever version we should use, that would at least include the Drag and Drop patch we obtained for macOS.

I have tried to understand what `EXULT_CHECK_SDL` does above `pkg-config` and I think this is :

1. Support `--with-sdl-prefix` and `--with-sdl-exec-prefix` in the `./configure` options, to get proper `SDL_CFLAGS` and `SDL_LIBS` variables. Is there any build ( macOS ? Android ? ) where SDL could be installed in an irregular location, such that its `sdl2.pc` would not be found, or such that the regular `sdl2.pc` would point to another SDL2 installation ?
2. Check for a valid `SDL.h` with the same version as `pkg-config`, setting a config variable `HAVE_SDL_H` that is nowhere used.
3. Check which variant of linkage works, that is without or with an explicit `SDL_main`, setting a Makefile variable `SDL_MAIN_NEEDED` that is nowhere used either.

Is there something that I have missed ? Or could we remove `acinclude.m4` too ?

**Comment by DominusExult** [comment]
There were good reasons back then and I don't remember these at all anymore.
First it was of course that we needed to have a proper way to choose between SDL 1.2x and SDL2.
That reason is now gone as we fully switched to SDL2. BUT SDL development has now tsken off to SDL3 so in the long term we might need to be able to choose again. Or not (SDL devs are actively developing a SDL2-compat wrapper to seamlessly transition and that might make easier for us OR when the time comes we just fully switch).
After the complete transition it was still a good idea to keep this m4 file around but really can't remember.

In the meantime I tried disabling SDL2 and OGG/Vorbis dependency for the Android build. But SDL2 is needed for building expack (for the data files) as the ./files/sdlrwops* sources depend on SDL2.
And, oddly, disabling the OGG/Vorbis dependency in configure made the Android build fail similar to when I tried building in on the Ubuntu-22.04 action runner image 🤷‍♂️

**Comment by Dragon-Baroque** [comment]
I have faced the same problem of a coexistence of and a choice between two versions of a product. 
For GTK with Exult Studio, initially GTK+ 2 and GTK+ 3, and now GTK+ 3 and GTK 4.
The problem has simplified because now the two versions can coexist, and because the two versions have separate `pkg-config` files.
In `configure.ac` it was only a matter of  a `--with-gtk` option taking values `gtk2`, `gtk3` or `gtk4`, and then of a `pkg-config` against GTK according to the `with_gtk` value like this, with a default of `gtk3` :
```
if test x$with_gtk = xgtk2; then
        PKG_CHECK_MODULES(GTK, [gtk+-2.0 >= 2.24], have_gtk=yes, have_gtk=no)
elif test x$with_gtk = xgtk4; then
        PKG_CHECK_MODULES(GTK, [gtk4            ], have_gtk=yes, have_gtk=no)
else
        PKG_CHECK_MODULES(GTK, [gtk+-3.0 >= 3.16], have_gtk=yes, have_gtk=no)
fi
```
I am quite sure that the same coexistence and `pkg-config` independence will occur with SDL2 and SDL3.

> But SDL2 is needed for building expack.

The rest of Exult for Android is a 4-way cross compilation ( ARM 32 and 64, X86 32 and 64 ), whereas `expack` is needed only once as X86 32 or 64. Do you think that either the data files could be built as a prior step of the GitHub workflow and be integrated when the APK is built, or the `expack` tool could be built as a prior step of the GitHub workflow and be used to build the data files in the APK build step ?

> The "latest" uses 22.04 and either that Ubuntu version or something that is installed crashes compilation without a clear indication what is going wrong.
> Curiously my Ubuntu VMs of 22.04.2 and 22.10 work correctly (with all the fixes mentioned here).

> Gradle needs to be <v7 (currently 6.9.4)
> Gradle 7.x also works.

In the commit 564331c5 which made the `ci-android` workflow work, you changed both the Ubuntu version from latest to 2020.04 and a whole build of Exult by a build limited to of `files/` and `data/`. Where both required or might the build succeed with only the restriction of the build set while remaining under Ubuntu latest ?

All `ci-android` build crashes I have looked at look like truncated recoveries by Java - `gradle` I guess - of an exception in a subprocess  - often the cross compiler. This exception is followed by an advice by `gradle` again, I suppose, of using `--stacktrace` and/or `--debug` to get more log. Now it looks like rather easy to perform this with `gradle` 7.4, using the 
`gradle.properties` lines `org.gradle.logging.stacktrace=all` and `org.gradle.debug=true`. With `gradle` 6.9, `org.gradle.logging.stacktrace=all` would not be permitted, that is why I am interested in your success with `gradle` 7.x.

> And, oddly, disabling the OGG/Vorbis dependency in configure made the Android build The "latest" uses 22.04 and either that Ubuntu version or something that is installed crashes compilation without a clear indication what is going wrong fail similar to when I tried building in on the Ubuntu-22.04 action runner image.

Is this failure without the OGG/Vorbis dependency visible as an action on GitHub or is it a test you made on your workstation ? I cannot find a failing `ci-android` action on GitHub after it succeeded by that commit.

**Comment by DominusExult** [comment]
I am using my fork to abuse Github's build actions. My trial withoutthe ogg/vorbis header files and disabling in configure.ac is at https://github.com/DominusExult/exultAndroCI/actions/runs/4489076692
Al these commits over there are not really something to look at but were used to limit the building on our upstream Exult.
But the more valuable your gradle property to trace the error. I'll give that a try in my playground tomorrow.
I wonder though, it seemed to me that the failing was always (?) in ./files with the sdlwrops* sources 🤷‍♂️


The reason I limited the workflow to build only a few folders is because all that is needed for Exult to build the APK are the data files (which needs ./files to build expack in ./tools). If I just run make in the root (of the out of source folder) it will build some more unnecessary before going into ./android and then afterwards, too.
That the data file generation needs SDL is not a big problem as it builds out of source and once the APK building starts that build system using gradle does not use the out of source folder and downloads SDL2, ogg/vorbis and mt32emu. I only wanted to make the build time just a bit slimmer. That's why it surprised me that the missing ogg/vorbis dev files and configure lines had any influence on the APK build.

**Comment by Dragon-Baroque** [comment]
At this point, I am interested in the nesting of the processes when the build crashes. Take the build
https://github.com/exult/exult/actions/runs/4466704272/jobs/7845223889
which is the last failing `ci-android` build before @DominusExult made it work.

Is the following nesting correct ?

- The GitHub workflow is running the `ci-android` ( `.github/workflows/ci-android.yml` ) Build step under Ubuntu 2022.04.2. The Build step has reached command `make -j2` in folder `android/`.
- `make` is running `./gradlew assembledebug` from `android/Makefile(.am)`.
- `gradle` is running `cmake --build` over `android/app/src/main/cpp/CMakeLists.txt` as part of Task `:app:generateJsonModelDebug`. I have no idea where it finds the set of Tasks.
- `cmake` is at `android/app/src/main/cpp/dependencies/exult/CMakeLists.txt:2` [ `android/app/src/main/cpp/dependencies/build_dependency.cmake:21` ] doing a nested `cmake` for Exult in `armeabi-v7a` ( ARM 32 ) mode. ARM 32 is the first mode of the `:app:generateJsonModelDebug` Task. Before reaching this point, the dependent ARM 32 builds for Ogg, Vorbis, mt32emu, and SDL2 have passed successfully.
- `cmake` ( more exactly `ninja`, I guess ) is running `libtool compile` over either Exult source `conf/XMLEntity.cc` or the next file in the build sequence, as the crash stdout line is truncated.
- `libtool compile` crashes but the stdout / stderr output of the crash is missing.

Trouble is, the stdout output is truncated, hopefully it is only the last line that gets truncated. This is classically because the stdout is buffered - for performance. In addition, no stderr output shows in the log.

@ceckak, @DominusExult, is there a way in the Gradle configuration, to force the output of stderr, and the non buffering of stdout - for the price of less performance and more verbosity ?

@DominusExult, I have looked at various failing `ci-android` logs going back in time from the one listed above, they all crash in the ARM 32 phase of Exult, but not always at the same place. The two immediate runs to the above fails at `shapes/fontvga.cc`, the prior one at `shapes/miscinf.cc`, then at `shapes/fontgen.cc`...

Now, have you tried to remove the `-j2` of the build of `android/` - I do not trust very much Java being parallelized ? Have you tried to increase the JVM memory size in `android/gradle.properties` ?

> It ends up using the Gradle and Java that you have in your path rather than the ones bundled with Android Studio.

@ceckak, you may be on something there. I am confused at the relationship between Gradle and Gradle Wrapper, considering that the Wrapper should essentially solve the coexistence issue. In the past, when I fought Java ( I hate Java ), I always ended up running _any_ Java applications with a whole explicit path, _and_ with JAVA_HOME and CLASSPATH set. Should the `android/Makefile(.am)` calls to `gradle` and `gradlew` be secured along these lines ?

**Comment by DominusExult** [comment]
I've set up a new fork and logged some sensible changes.
- got properly rid of host configure dependencies and set exult make to run single threaded and with the quiet arg (which doesn't silence much) **(still using Ubuntu 20.04, the build runs fine)**
- switched to Ubuntu-latest (22.04) and the build crashes as usual
- added stacktrace=all logging (of course still crashes)
you can see the action logs at https://github.com/DominusExult/exult/actions
These all have their last logged build action ./shapes/shapeinf so maybe there is something going on in there OR it's just a coincidence that it breaks there.

@Dragon-Baroque I've added you as a collaborator to that fork. If you have any idea to proceed, just push an update. Maybe in a branch, though, which I should have done to begin with in this fork.

**Comment by Dragon-Baroque** [comment]
Thank you @DominusExult.

Before I push any update, I intend to exploit the run attempts you made, looking at the differences in the logs between job 2 ( ubuntu 20.04 ) and job 3 ( identical to job 2 except switched to ubuntu latest ).

For me to be able to push update on your fork, I suppose I have to clone it at home, correct ?

Already, by comparing the Build log of jobs 2 and 3, I have come to something very fishy, that is that someone, I suspect this is Gradle, bufferizes and limits and truncates stdout. **Also in job 2 which passes**. See for example that the log of `debug|armeabi-v7a` ( ARM 32 ) in job 2 stops at the same place as in job 3, that the other parts `debug|arm64-v8a` ( ARM 64 ), `debug|x86` ( X86 32 ) and `debug|x86_64` ( X86 64 ) stop at various places in the sources and that none log the linkedit phase, producing `exult` or `libexult.so`.

In other words, we cannot trust the run log to understand where and how it failed.

Thanks for the `stacktrace=all` but indeed it does not show anything worthwhile.

**Comment by DominusExult** [comment]
yes, you need to clone my fork and then just push.
I've experimented in https://github.com/DominusExult/exult/commit/3ca70ec62f7ffc785983677a28fa6c505aebc9bb to see if it indeed crashes in ./shapes which it does. I've run the job twice. The first fail seemed to have problems in ./shapes/shapeinf but then finished it on make in ./shapes where it failed again.
The second time I've run that job it finished in ./shapes/spapeinf but again failed in ./shapes but the last file it completed was different to the first run.
So something really odd is happening there.

**Comment by Dragon-Baroque** [comment]
> Yes, you need to clone my fork and then just push.

Not permitted, I tried both a new branch and a small commit, and then that same small commit to master. Both failed :
```
To https://github.com/DominusExult/exult.git
 ! [remote rejected]     baroque -> baroque (permission denied)
error: failed to push some refs to 'https://github.com/DominusExult/exult.git'
```
and same with `master -> master`.
And yes, I have switched the clone user to Dragon-Baroque.

I was trying to silent further the Exult build under Gradle replacing in `android/app/src/main/cpp/dependencies/exult/CMakeLists.txt.in` : 
`BUILD_COMMAND     make -s` 
by 
`BUILD_COMMAND     make -s LIBTOOLFLAGS="--silent"`
so that, hopefully, the log truncation would not occur before the build failure in Ubuntu latest and the build failure would show up in the log.

**Comment by DominusExult** [comment]
@Dragon-Baroque I just noticed one difference between the good 20.04 run and the bad 22.04 run:
seehttps://github.com/DominusExult/exult/actions/runs/4504543486/jobs/7929149724#step:8:10768 <-bad
https://github.com/DominusExult/exult/actions/runs/4503315568/jobs/7926343661#step:8:10889 <- good
`[6/8] Performing build step for 'exult-1.9.0git'` seems to fail on the 22.04 runner, I think on all failed runs. 
```
/home/runner/work/exult/build/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : FAILED: exult-1.9.0git-prefix/src/exult-1.9.0git-stamp/exult-1.9.0git-build 
[10769](https://github.com/DominusExult/exult/actions/runs/4504543486/jobs/7929149724#step:8:10770)

[10770](https://github.com/DominusExult/exult/actions/runs/4504543486/jobs/7929149724#step:8:10771)
/home/runner/work/exult/build/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : FAILED: exult-1.9.0git-prefix/src/exult-1.9.0git-stamp/exult-1.9.0git-build 
[10771](https://github.com/DominusExult/exult/actions/runs/4504543486/jobs/7929149724#step:8:10772)
/home/runner/work/exult/build/android/app/src/main/cpp/CMakeLists.txt : C/C++ debug|armeabi-v7a : cd /home/runner/work/exult/build/android/app/.cxx/cmake/debug/armeabi-v7a/dependencies/exult/exult-1.9.0git-prefix/src/exult-1.9.0git-build && make -s && /usr/local/lib/android/sdk/cmake/3.18.1/bin/cmake -E touch /home/runner/work/exult/build/android/app/.cxx/cmake/debug/armeabi-v7a/dependencies/exult/exult-1.9.0git-prefix/src/exult-1.9.0git-stamp/exult-1.9.0git-build

```

Now what is that build step?

**Comment by Dragon-Baroque** [comment]
@DominusExult I may have lost the invite email. But I have found the invite in my GitHub user profile. Now accepted.

> Now where is that happening ?

The Build log under Gradle is hard to read because it is not chronological. Gradle submits the C/C++ compilation and link to CMake, using a separate process `cmake-server` and a communication API to retrieve the completion code and the stdout / stderr from the CMake server. Gradle launches a `cmake-server` process for each target build ( ARM 32 and 64, X86 32 and 64 ). That process builds all dependencies for the requested target build ( Ogg, Vorbis, MT32Emu, SDL2 and finally Exult ). 

_Gradle displays the CMake target build completion first, that is success or failure, and the CMake retrieved stdout / stderr log only afterwards._ So that the answer to your question is : look just below in the retrieved CMake log.

Now there is also a second problem in this architecture. The retrieved CMake stdout / stderr log is somehow size limited and **truncated** if its size is larger than the size limit. In our case, the log was truncated _before_  the compilation failure, making it invisible. I cannot pinpoint the culprit, among CMake Server, Gradle and the API between them, but the solution was clear, reduce the size of the output log enough to pass below the size limit.

This was why I tried the build Exult with `make -s LIBTOOLFLAGS="--silent"` which is more silent than just `make -s`.
And you obtained the complete log with the build failure.

By the way, there is a third problem, as each retrieved stdout / stderr from CMake Server is displayed _twice_, with an empty line in between. Please look at the log ...

> Failing in expack but I don't understand why it even bothers building expack as both tools and data building is disabled.

Now **that is the** question. I would like

1. to revert your patch isolating the compilation of ./shapes, and rerun the Build, checking that it still fails upon `expack`,
2. to try the same under Ubuntu 20.04, please note that the log truncation occurs also, and see whether it tries to build `expack`.

**Comment by Dragon-Baroque** [comment]
@DominusExult, I restored your branch `going_through_make` to Ubuntu latest, and I removed the `stacktrace=all`.

Now it differs from `exult/exult` at `configure.ac`, `files/Makefile.am`, and at the various `make -s LIBTOOLFLAGS="silent"`.

Of course, reverted to Ubuntu latest, it fails again at `tools/expack`, this is `ci-android #9` run https://github.com/DominusExult/exult/actions/runs/4518410743.
I leave it to you to decide whether the Android build _should_ build `expack`.

**Comment by DominusExult** [comment]
Thanks. Odd that we still build expack even though building data is disabled. Seems like this is an oversight. Midisfx.flx is stored as binary in our code, so that doesn't need building anyway.

The big question is still why building expack fails on the 22.04 runner while it works in my 22.04 VM. *Something* that is installed or maybe missing is interfering.

**Comment by Dragon-Baroque** [comment]
> The big question is still why building expack fails on the 22.04 runner while it works in my 22.04 VM.

The `expack.cc` includes `unistd.h` then some Android SDK / NDK provided files that end up in `features.h`.
`features.h` on its turn includes `sys/cdefs.h` _which, on the 2022.04 runner, is not provided by the Android SDK / NDK_ but by the host glibc instead. If you look at the file paths in the failing build, you notice the jump from the Android SDK/NDK to a regular `/usr/include` folders.

The build fails on a `__GNUC_PREREQ` macro in `sys/cdefs.h` because the Android SDK / NDK Clang compiler at version 9.0.9 does not provide this macro.

There could be differences either in the VM Ubuntu 22.04 glibc package or in the downloaded Android SDK / NDK. Perhaps a newer CLang would provide the macro. Perhaps a different glibc would front this macro with some check for existence.  I don't know. A build log of Android Exult from your 22.04 VM might help.

By the way, on what criteria do you choose and where do you specify the _required version_ of Android Studio or SDK / NDK ? I have not been able to locate it. How do you choose the version of Gradle and of Java, also ?

The reason is that I have been frustrated by the use of the CMake Server by the 22.04 runner, which is deprecated since CMake 3.15 and has been removed from CMake 3.20. The CMake in my Linux is at 3.26. The CMake Server is replaced by a CMake File API which is supposed to be supported by Android Studio 4.2.

**Comment by Dragon-Baroque** [comment]
> Odd that we still build expack even though building data is disabled.

The configure options `--disable-tools` and `--disable-data` only set two Automake variables to false, respectively `BUILD_TOOLS` and `DATA_FILES`. They do not ( and cannot ) prevent the navigation of the `tools/` and `data/` folders during the build. The `Makefile.am` of these two folders exploit the Automake variables in `if` sections, but `expack` and `midisfx.flx` are out of the `if` sections, thus they are built unconditionally.

I looked at the commit backlog and I spotted this :
```
commit 60f90f4394c5cce7715d2532393a3386f500d315
Author:     Willem Jan Palenstijn <wjpalenstijn@users.sourceforge.net>
AuthorDate: Thu Aug 9 12:51:08 2001 +0000
Commit:     Willem Jan Palenstijn <wjpalenstijn@users.sourceforge.net>
CommitDate: Thu Aug 9 12:51:08 2001 +0000

    * tools/Makefile.am: always build expack, even with --disable-tools
```
```
 if BUILD_TOOLS
-noinst_PROGRAMS = expack wuc wud mklink rip ucdump shp2pcx splitshp $(UCC)
+EXTRATOOLS = wuc wud mklink rip ucdump shp2pcx splitshp $(UCC)
+else
+EXTRATOOLS =
 endif
 
+noinst_PROGRAMS = expack $(EXTRATOOLS)
+

```
which takes `expack` out of the `if BUILD_TOOLS`.

**Comment by Dragon-Baroque** [comment]
> Odd that we still build expack even though building data is disabled.

By the way, there as some code in `configure.ac` that handles `expack` and `head2data` specially under a `cross_compiling` variable that I cannot find any other occurrence of. You wrote this in a commit circa 2013 as b4c19ea99.

**Comment by DominusExult** [comment]
I think autoconf/automake sets CROSS_COMPILING automatically when build and host do not match. I use that a lot when building the macOS snapshots as I‘m handling three different arches.

Hard to say what wjp's thoughts were to take expack out of this 22 years ago 🤷‍♂️

How do I check for glibc? Is that just plain libglib-xxxx?

I'll see about adding a log of my successful build on the VM. Probably a config.log would be as revealing 🤷‍♂️

**Comment by Dragon-Baroque** [comment]
> I think autoconf/automake sets CROSS_COMPILING automatically when build and host do not match.

Then it should apply to the Android builds as well. At least against ARM.

> How do I check for glibc? Is that just plain libglib-xxxx?

libc6, libc-dev-bin, linux-libc-dev.
glibc is the GNU libc whereas glib is the Graphical library at the root of GTK, Gnome, Gimp...

**Comment by DominusExult** [comment]
I'll try next how it behaves when I set the expack building into an if cross_compile

**Comment by DominusExult** [comment]
@Dragon-Baroque @marzojr: disabling building expack when cross compiling makes the build run through.
So even though it seems to be a bug somewhere in the runner image, I propose these extensive changes to Exult's build system:
https://github.com/DominusExult/exult/commit/b9e220129c430e56a8a257d325bb5639998c6ee9

- This eliminates all need for SDL2, ogg/vorbis dependencies to be installed when wanting to build the APK. This meant placing the Android APK tests way up in configure.ac, adding `AM_CONDITIONAL(SDL_MAIN_NEEDED, false)` (otherwise acinclude.m4 would complain - but ultimatively we don't even need this for anything, right) and building ./files/sdlrwops*.* only when building Exult.
- Not building Expack when cross compiling *and* disabling building tools (enable_tools=no should/could be added to the APK check in configure.ac)
- and just for readability of the build logs, silencing the building of Exult, ogg and vorbis (No idea how to silence the cmake (ninja?) building of SDL2 and MT32Emu)

Is this acceptable?

**Comment by Dragon-Baroque** [comment]
> Is this acceptable ?

Yes, considering my lack of knowledge of the Android environment.
My bottom line : Is the built APK installable and working ?

**Comment by DominusExult** [comment]
THAT is a question @ceckak needs to answer ;)
Our APK build has two targets, debug and release. The difference is that the release one needs to be signed. And as far as I understood @ceckak you need the signed one to install it on a normal Android mobile/tablet.
The debug built *might* be installable on the emulator but I don't know and need to test yet.

**Comment by ceckak** [comment]
If you can point me at a fork I can clone, I'll try to test the release build this weekend.  The debug build can be installed on an emulator or hardware device, but you can't switch back and forth between debug and release on the same device.

**Comment by DominusExult** [comment]
My fork https://github.com/DominusExult/exult and important, the branch all_fixes would be great to test. This should work and have all the updates.

If it works, I think the only part that could use fixing is that the sources for ogg, vorbis and mt32emu should also be downloaded only once, instead of for each of the 4 Android arches, similar to SDL2. Mostly because I had builds fail because, mysteriously, on the second or third arch, connection to github was down.

**Comment by DominusExult** [comment]
@ceckak the debug build won't run in the emulator. The launcher works and I was able to set up the games but then launching Exult immediately crashes :(

**Comment by DominusExult** [comment]
@ceckak I'm in my final step of testing but it seems *I* broke the Android port by upgrading the used SDL2 version without testing.
Anything > 2.0.22 breaks the Android port (and I'll make the necessary commit later to downgrade SDL2 again). But finding out what is causing it might be something for you to look at.
And I'll see about adding the APK to our snapshot creation workflow.

**Comment by ceckak** [comment]
Hoping to take a look at it this evening.  I had tried a build with an updated SDL and it worked on my tablet but ran really slow.  Haven't had a chance to debug it yet.

**Comment by DominusExult** [comment]
I may be wrong though, it rather looks as if the change I made in android/app/src/main/cpp/dependencies/sdl/CMakeLists.txt in 697a67674 makes it crash. I have no idea why, though. but I'm building again to see if it's that

**Comment by ceckak** [comment]
Building locally (on a mac), I'm promptly running into an issue with not finding SDL.h - probably related to some of your changes with SDL detection on the branch:

```
libtool: compile:  clang++ -std=c++17 -DHAVE_CONFIG_H -I. -I../../exult/imagewin -I.. -I../../exult/headers -I../../exult -I../../exult/files -DMACOSX -I../../exult/conf -Wall -Wextra -pedantic -Walloca -Wc++14-compat -Wc++17-compat -Wbool-operation -Wcast-align -Wcast-qual -Wcast-function-type -Wctor-dtor-privacy -Wdisabled-optimization -Wextra-semi -Wformat-nonliteral -Wformat-security -Wlogical-not-parentheses -Wmissing-include-dirs -Wnon-virtual-dtor -Wnull-dereference -Wold-style-cast -Woverloaded-virtual -Wredundant-decls -Wshift-negative-value -Wshift-overflow -Wtrigraphs -Wundef -Wuninitialized -Wwrite-strings -Wzero-as-null-pointer-constant -Wabsolute-value -Wdeprecated-register -Wmismatched-tags -Wunused-private-field -stdlib=libc++ -MT iwin8.lo -MD -MP -MF .deps/iwin8.Tpo -c ../../exult/imagewin/iwin8.cc -o iwin8.o
../../exult/imagewin/iwin8.cc:38:10: fatal error: 'SDL.h' file not found
#include <SDL.h>
         ^~~~~~~
1 error generated.
make[2]: *** [iwin8.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

Prior to that, it was successfully finding `sdl2-config` and using that to configure the paths.

As to the crash you were seeing from your changes, I suspect it was due to removing the vorbis dependency.  I dug back through the commits and don't see any specific notes on why I did that, but when I look [here](https://github.com/exult/exult/commit/46f069d815267cbecd09a7299130b4461bff0e6b#diff-184f1744549930f6ffdb572c2f4b2fcc89cfc880138326045425a0171e751101), I see that when I added the vorbis dependency to SDL, I also changed `android/app/src/main/cpp/CMakeLists.txt` to depend on `SDL2` *instead* of `vorbis`.  By removing the vorbis dependency from sdl, you effectively removed it from the dependency chain completely, so while it may get built, the library probably isn't getting bundled into the APK.

Why I made vorbis a dependency of SDL rather than pulling it in at the exult level I don't recall, but your change may work if you add the vorbis dependency in `android/app/src/main/cpp/CMakeLists.txt`.

**Comment by DominusExult** [comment]
ah yes, don't bother with make in the root of the build dir. Just
```
cd ./../build/files && make -j
cd ./../data && make -j
cd ./../android && make -j
```

(but I need to fix the imagewin make)

**Comment by Dragon-Baroque** [comment]
> SDL anything > 2.0.22 breaks the Android port.

Some higher version of the Android SDK to request, perhaps ?
You have `ANDROID_SDK_VERSION=30` in `configure.ac` for `android/Makefile.am`.
Whereas the `docs/README-android.md` of SDL2 2.26 states a requirement of `Android SDK (version 31 or later)`.

**Comment by DominusExult** [comment]
> > SDL anything > 2.0.22 breaks the Android port.
> 
> Some higher version of the Android SDK to request, perhaps ? You have `ANDROID_SDK_VERSION=30` in `configure.ac` for `android/Makefile.am`. Whereas the `docs/README-android.md` of SDL2 2.26 states a requirement of `Android SDK (version 31 or later)`.

Sorry, that was not clear from my posts, it actually was not an SDL2 version problem but 

> As to the crash you were seeing from your changes, I suspect it was due to removing the vorbis dependency. I dug back through the commits and don't see any specific notes on why I did that, but when I look [here](https://github.com/exult/exult/commit/46f069d815267cbecd09a7299130b4461bff0e6b#diff-184f1744549930f6ffdb572c2f4b2fcc89cfc880138326045425a0171e751101), I see that when I added the vorbis dependency to SDL, I also changed android/app/src/main/cpp/CMakeLists.txt to depend on SDL2 instead of vorbis. By removing the vorbis dependency from sdl, you effectively removed it from the dependency chain completely, so while it may get built, the library probably isn't getting bundled into the APK.
> 
> Why I made vorbis a dependency of SDL rather than pulling it in at the exult level I don't recall, but your change may work if you add the vorbis dependency in android/app/src/main/cpp/CMakeLists.txt.
I fixed this in 3d6c8224277500247f6345e5ff420970108af1c5

**Comment by DominusExult** [comment]
@ceckak I've now committed everything to main Exult. Having no SDL2 (or not having the detection of it) should no longer be a problem. Building the APK on macOS also had some weird png.h detection problem which I circumvented by forcefully not detecting it for a ARCH=Android.
So building the APK works now on the GitHub runner image 22.04 as well and if you want to enhance the APK you can do it without first worrying about the build working or not :)

**Comment by DominusExult** [comment]
Yes, please. I was only able to test on the emulator and only tested with one "device" and only to see it launching and hear the menu music.

**Comment by ceckak** [comment]
Finally was able to test on my tablet.  I'm seeing the same issue with the current codebase that I did previously trying to update to the latest SDL - it runs unusably slow.  Could be related to updating the sdk version as @Dragon-Baroque suggested above, but I haven't experimented with it yet.  In the mean time, I'd suggest changing back to the older SDL version:

```
diff --git a/android/Makefile.am b/android/Makefile.am
index fc09bb7d7..6aeacaa21 100644
--- a/android/Makefile.am
+++ b/android/Makefile.am
@@ -116,4 +116,4 @@ $(APK_ASSETS): $(APK_ASSETS_DIR)/%: $(abs_top_srcdir)/data/%
 # The NDK build also utilizes this codebase rather than downloading a
 # separate copy.
 $(SDL_SOURCE_DIR):
-       $(GIT) -c advice.detachedHead=false clone --branch release-2.26.4 --depth 1 https://github.com/libsdl-org/SDL.git $@
+       $(GIT) -c advice.detachedHead=false clone --branch release-2.0.22 --depth 1 https://github.com/libsdl-org/SDL.git $@
```

Also, I noticed with this build that when I open my inventory, there is a "usecode container" bag thingy which has a bunch of numbers inside it.  When I go back to a previous build, that doesn't show up.  Guessing that is not related to Android.  When I get time, I'll see if I can replicate it with a desktop build and post a separate issue for it.

**Comment by Dragon-Baroque** [comment]
> Also, I noticed with this build that when I open my inventory, there is a "usecode container" bag thingy which has a bunch of numbers inside it.

@DominusExult, could this come from your commit bd0a8b25 of March 19, which altered the `configure` command in
`android/app/src/main/cpp/dependencies/exult/CMakeLists.txt.in` :

```
-  COMMAND           . @ENVFILE@ && <SOURCE_DIR>/configure --prefix=<INSTALL_DIR> --host=@ANDROID_LLVM_TRIPLE@ --enable-libexult --enable-shared --disable-data --disable-tools --enable-mt32emu --disable-fluidsynth
+  COMMAND           . @ENVFILE@ && <SOURCE_DIR>/configure --prefix=<INSTALL_DIR> --host=@ANDROID_LLVM_TRIPLE@ --enable-libexult --enable-shared --disable-data --disable-tools --enable-zip-support --enable-midi-sfx --enable-mt32emu --disable-fluidsynth --disable-alsa --disable-timidity-midi --disable-oggtest --disable-vorbistest --enable-usecode-container --enable-nonreadied-objects
```
where there is now a `--enable-usecode-container` ?

**Comment by DominusExult** [comment]
Yes, that is added through this configure args.
Don't use these for a real buuld (unless you want to debug the games).
They are only added for the CI to see if it builds with these.

**Comment by ceckak** [comment]
Is that something you were just testing temporarily, or are you wanting to keep it enabled in all of the CI builds?  If the latter, maybe we can come up with a way to pass it in from the original top-level configure so that it isn't enabled by default in android builds.

**Comment by DominusExult** [comment]
The CI build is purely for testing whether new code builds. So it makes no difference.
But for snapshots like the Windows one these configure args would not be used.

**Comment by ceckak** [comment]
That makes sense, but the android build always uses all the flags from [here](https://github.com/exult/exult/blob/master/android/app/src/main/cpp/dependencies/exult/CMakeLists.txt.in#L11), irrespective of whether it is CI, snapshot, or someone manually cloning the github repo and building locally.  We probably want to trim the default flags in `CMakeLists.txt.in` to what we want in a snapshot/release, and add a way to pass additional flags in for a CI build.  Or find some other way to conditionally add the extra flags for CI.

**Comment by DominusExult** [comment]
D‘oh! I wasn‘t seeing what you mean 🤦‍♂️🤦‍♂️🤦‍♂️
